### DISCUSSION TITLE
Simplified effect cancellation for a subview

### DISCUSSION THREAD

**Discussion by surajbarkale** [pointfreeco/swift-composable-architecture#891] [Show and tell]
There is already a lot of discussion around how to handle effect cancellation when a view is dismissed. @darrarski has kindly created a [summary of all approaches](https://github.com/darrarski/darrarski/blob/main/2021/04/Thoughts-on-SwiftUI-navigation/README.md#dismissing-screen-and-canceling-its-side-effects). The minimum effort implementation is the [`.presents()` extension](https://github.com/pointfreeco/swift-composable-architecture/blob/9ec4b71e5a84f448dedb063a21673e4696ce135f/Sources/ComposableArchitecture/Reducer.swift#L549-L572) on `Reducer`.

However, I have found the `.presents()` function hard to compose when using deeply nested states (e.g. Tagged). So I have updated it as an overload on `.optional()`:

```swift
extension Reducer {
  public func optional(cancelEffectsOnNil: Bool) -> Reducer<State?, Action, Environment>  {
    let reducer = self.optional()
    guard cancelEffectsOnNil else {
      return reducer
    }
    let id = UUID()
    return .init({ state, action, environment in
      let hadState = state != nil
      let effects: Effect<Action, Never>
      if hadState {
        effects = reducer
          .run(&state, action, environment)
          .cancellable(id: id)
      } else {
        effects = .none
      }
      let hasState = state != nil
      return .merge(effects, hadState && !hasState ? .cancel(id: id) : .none)
    })
  }
}
```

With this extension, simply changing `.optional()` to `.optional(cancelEffectsOnNil: true)` will cancel all effects of the child state when it becomes nil.

**Comment by darrarski** [comment]
Hi @surajbarkale! I tried to do it this way too, but it didn't fit some of my use cases. 

- It's usually the parent who presents and dismisses the child. For that reason, state mutation connected to presentation happens in a parent reducer. We need to handle the case when the presented state becomes `nil` to cancel effects. That's why my `.presenting` function on reducer actually combines two reducers, makes sure they are called in the correct order, and that effects are canceled on dismiss.
- For performance reasons, I am trying to limit the number of produced `Effect.cancellable(id:)` and  `Effect.cancel(id:)` effects. It's actually quite simple and brings noticeable improvements.

Please check out [Swift Composable Presentation](https://github.com/darrarski/swift-composable-presentation) library.

Could you provide more details about what problems did you experience when using deeply nested states (e.g. Tagged)?

**Reply by surajbarkale** [reply]
Thanks! I shall take a detailed look at your library. It looks like something that should be integrated within TCA. My solution above is good enough for very simple use cases similar to the original `.presents()` function.

This is the view arrangement I am working with. There is an inline editor (`ChildView`) which has a long running effect (not shown in the example). I don't have enough brainpower to use `.presents` in this case.

```swift
struct ChildView: View {
  struct State: Equatable {
    @BindableState var value: String
  }

  enum Action: Equatable, BindableAction {
    case binding(BindingAction<State>)
    case save
  }

  struct Environment { }

  static let reducer = Reducer<State, Action, Environment> { state, action, environment in
    switch action {
    case .binding:
      return .none
    case .save:
      return .none
    }
  }.binding()

  let store: Store<State, Action>

  var body: some View {
    WithViewStore(store) { viewStore in
      VStack(alignment: .leading) {
        Text("Editor")
        TextField("Edit", text: viewStore.binding(\.$value))
          .textFieldStyle(.roundedBorder)
        Button("Save") { viewStore.send(.save) }
      }
      .padding()
    }
  }
}

struct ContainerView: View {
  typealias ChildID = UUID
  struct State: Equatable {
    var children: IdentifiedArrayOf<Identified<ChildID, String>>
    var editor: Identified<ChildID, ChildView.State>?
  }

  enum Action: Equatable {
    case edit(ChildID)
    case editor(ChildView.Action)
  }

  struct Environment { }

  static let reducer = Reducer<State, Action, Environment>.combine(
    ChildView.reducer
      .pullback(state: \Identified.value, action: .self, environment: { $0 })
      .optional()
      .pullback(state: \.editor, action: /Action.editor, environment: { _ in .init() }),
    Reducer { state, action, environment in
      switch action {
      case let .edit(key):
        if let child = state.children[id: key] {
          state.editor = .init(.init(value: child.value), id: child.id)
        }
        return .none
      case .editor(.save):
        if let editor = state.editor {
          state.children[id: editor.id]?.value = editor.value.value
        }
        state.editor = nil
        return .none
      case .editor:
        return .none
      }
    }
  )

  let store: Store<State, Action>

  var body: some View {
    WithViewStore(store) { viewStore in
      VStack {
        IfLetStore(store.scope(state: { $0.editor?.value }, action: Action.editor)) {
          ChildView(store: $0)
          Divider()
        }
        List {
          ForEach(viewStore.children) { child in
            HStack {
              Text(child.value)
              Spacer()
              Button("Edit") { viewStore.send(.edit(child.id)) }
            }
          }
        }
      }
    }
  }
}

struct ComtainerView_Previews: PreviewProvider {
  static var previews: some View {
    ContainerView(store: .init(
      initialState: .init(children: [
        .init("A", id: .init()),
        .init("B", id: .init()),
        .init("C", id: .init()),
      ]),
      reducer: ContainerView.reducer,
      environment: .init()
    ))
  }
}
```

**Reply by darrarski** [reply]
I think you can use `ForEachStore` to achieve it. Please check out this example:

> ### ➡️ [ForEachStore example](https://github.com/darrarski/swift-composable-presentation/blob/main/Example/Example/ForEachStoreExample.swift)
> 
> SwiftUI example with a list of components. When a component is deleted from the list, all effects produced by its reducer are canceled.

Let me know if it helps.

**Reply by surajbarkale** [reply]
I explicitly don't want all the child states to be stored in the container. I did not implement it in the first place because some of the child state needs to be loaded when the editor view is presented.
